consensus/istanbul/backend/announce.md

```diff
@@ -0,0 +1,217 @@
+
+As part of the protocol's design, a [NearlyElectedValidator] can advertise different [eNodeURL] values for different destinations, allowing for multi-proxy implementations of the protocol. That being said, the announce protocol itself is agnostic to the concept of proxies, it cares only for the sharing of `<validator address, validator eNodeUrl>` tuples. It is the proxy implementation's responsibility to ensure the correct behavior of this specification.
+
+### Concealing eNodeURL values
+
+It is important to ensure that [eNodeURL] values can't be read by nodes that are not in the [NearlyElectedValidator] set. To achieve this, they are shared in one of two ways:
+
+* Sent through a direct p2p connection from one [NearlyElectedValidator] to another ( [enodeCertificateMsg] )
+* Gossipped through the network, but encrypted with the public key of the recipient [NearlyElectedValidator] ( [queryEnodeMsg] )
+
+
+### Minimizing network traffic
+
+In order to reduce the amount of message flooding, every [FullNode] participating in the p2p network must maintain a version table for the [NearlyElectedValidator] nodes in the network. This table is built and updated with one specific message from the protocol ( [versionCertificatesMsg] ).
+
+[NearlyElectedValidator] can also use this table to understand which [eNodeURL] entries are stale, if the [eNodeURL] table entry has a version older than the version table entry.
+
+## Basic Operation
+
+When a validator is close to being elected ([NearlyElectedValidator]), it starts periodically sending [queryEnodeMsg] messages through the p2p network. These message are regossipped by nodes, and validators reply with a direct message to the originator with an [eNodeCertificateMsg], holding their `eNodeURL`. The initial [queryEnodeMsg] contained the origin validator's `eNodeURL` encrypted with the destination public key, therefore after the direct reply, both validators are aware of each others' `eNodeURL`.
+
+### p2p connection management
+
+Wether a validator opens a connection or not against another validator is not in the scope of this spec, but only how and when should the protocol messages should be sent, replied and regossipped. For example, it IS part of the spec what messages should be sent to a newly connected peer or validator peer.
+
+## Protocol Messages (RLP)
+
+All messages presented here are Istanbul messages, following the same format as the consensus messages (see `istanbul.Message`), which are serialized in the `payload` field for standard p2p messages.
+
+`[code: P, msg: B, sender_address: B_20, signature: B]`
+
+- `code`: the message code (0x12, 0x16, or 0x17).
+- `msg`: rlp encoded data, depending on the message type (`code`), and detailed in the following subsections.
+- `sender_address`: the originator of the message.
+- `signature`: the message signature.
```

Comment:
Currently there doesn't seem to be any reference to the `sender_address` in the rest of the spec, and I think the sender address is critical to specifying the protocol, because it is the sender address that lets nodes construct the mapping between address and eNodeURL.

Further down in the message handling descriptions talk about upserting the eNodeURL, but they don't mention the key used to upsert, which is the sender_address.

Reply:
sender address is a parameter from the 'parent' type istanbul.Message. That's why it's only being mentioned as to where it's coming from, and it's assumed the istanbul message format is known